src/ruis/render/scene/gltf_loader.cpp

```diff
@@ -0,0 +1,772 @@
+
+	const uint32_t buffer_view_index = read_uint(accessor_json, "bufferView");
+	const uint32_t acc_count = read_uint(accessor_json, "count");
+	const uint32_t acc_offset = read_uint(accessor_json, "byteOffset");
+	const uint32_t component_type = read_uint(accessor_json, "componentType");
+
+	auto new_accessor = utki::make_shared<accessor>(
+		buffer_views[buffer_view_index],
+		acc_count,
+		acc_offset,
+		// acc_stride,
+		type_v,
+		static_cast<accessor::component_type>(component_type)
+	);
+
+	const uint32_t bv_offset = new_accessor.get().bv.get().byte_offset;
+	const uint32_t bv_length = new_accessor.get().bv.get().byte_length;
+	const uint32_t bv_stride = new_accessor.get().bv.get().byte_stride;
+
+	auto buf = glb_binary_buffer.subspan(bv_offset + acc_offset, bv_length);
+
+	// const size_t num_components = static_cast<const size_t>(new_accessor.get().type_);
+
+	// if (new_accessor.get().bv.get().target_ == buffer_view::target::array_buffer) {
+
+	if (new_accessor.get().component_type_v == accessor::component_type::act_float) {
+		if (new_accessor.get().type_v == accessor::type::scalar)
+			create_vertex_buffer_float<float>(new_accessor, buf, acc_count, bv_stride);
+		else if (new_accessor.get().type_v == accessor::type::vec2)
+			create_vertex_buffer_float<ruis::vec2>(new_accessor, buf, acc_count, bv_stride);
+		else if (new_accessor.get().type_v == accessor::type::vec3)
+			create_vertex_buffer_float<ruis::vec3>(new_accessor, buf, acc_count, bv_stride);
+		else if (new_accessor.get().type_v == accessor::type::vec4)
+			create_vertex_buffer_float<ruis::vec4>(new_accessor, buf, acc_count, bv_stride);
+		else {
+			throw std::logic_error("Matrix vertex attributes currently not supported");
+		}
+
+		// create_vertex_buffer_float<r4::vector<float, num_components>>(buf);
+	} else if ((new_accessor.get().component_type_v == accessor::component_type::act_unsigned_short ||
+				new_accessor.get().component_type_v == accessor::component_type::act_unsigned_int) &&
+			   // new_accessor.get().bv.get().target_ == buffer_view::target::element_Array_buffer &&
+			   // there are gltf exporters which don't mark target at all. So we must detect index arrays another way
+			   new_accessor.get().type_v == accessor::type::scalar)
+	{
+		utki::deserializer d(buf);
+		if (new_accessor.get().component_type_v ==
+			accessor::component_type::act_unsigned_short /*|| acc_count < 65536*/)
+		{
+			std::vector<uint16_t> index_attribute_buffer;
+			index_attribute_buffer.reserve(acc_count);
+
+			for (size_t i = 0; i < acc_count; ++i)
+				index_attribute_buffer.push_back(d.read_uint16_le());
+
+			new_accessor.get().data = index_attribute_buffer;
+			new_accessor.get().ibo = factory_v.create_index_buffer(utki::make_span(index_attribute_buffer));
+
+		} else if (new_accessor.get().component_type_v ==
+				   accessor::component_type::act_unsigned_int /*|| acc_count >= 65536*/)
+		{
+			std::vector<uint32_t> index_attribute_buffer;
```

Comment:
add `// TODO: create 16 bit index buffer if less than 65536 vertices`

Reply:
16 bit indices are already supported, but instead of checking the actual number of vertices/indices we check the data type from GLTF file : new_accessor.get().component_type_v == accessor::component_type::act_unsigned_short

Reply:
That's right, but what I mean is the optimization: in case GLTF says that index type is 32 bit, but still provides less than 65536 vertices, then there is no reason to use 32 bit index, we can convert it to 16 bit index. It's just an optimization which is not of a priority to implement, so I thought of adding a TODO about it.